src/main/java/com/dougnoel/sentinel/pages/PageManager.java

```diff
@@ -165,88 +158,132 @@ public Page maximizeWindow() {
+     * @param pageName The name of the page to switch to
+     * @return The window handle of the switched-to window
+     * @throws NoSuchWindowException If no page of that name was utilized previously
      */
-	public static String switchToNewWindow(String pageName) throws InterruptedException {
-		String newHandle = null;
-		parentHandle = driver().getWindowHandle();
-		parentPage = PageManager.getPage().getName();
-		Set<String> handles = driver().getWindowHandles();
-		if (handles.size() == 1) {
-			var errorMessage = "Only one window is open, therefore we cannot switch to a new window. Please open a new window and try again.";
-			log.error(errorMessage);
-			throw new NoSuchWindowException(errorMessage);
+    public static String switchToExistingWindow(String pageName) {
+    	previousPageInfo = currentPageInfo;
+
+    	if(pages.containsKey(pageName)) {
+    		driver().switchTo().window(pages.get(pageName).getRight()[0]);
+    		setPage(pageName);
+    	}
+    	else {
+    		throw new NoSuchWindowException("No page of that name was used previously");
+    	}
+
+    	return driver().getWindowHandle();
+    }
+
+    /**
+     * Switches to the first found unvisited page and associates it with the page object
+     * associated with the passed page name.
+     * <p>
+     * <b>Preconditions:</b> Only one new window should have opened.
+     * @param pageName The page object to associate with the newly switched to page
+     * @return The window handle of the switched-to window
+     * @throws NoSuchWindowException if a new window cannot be switched to in the configured
+     * timeout.
+     */
+    public static String switchToNewWindow(String pageName) {
+    	previousPageInfo = currentPageInfo;
+    	ArrayList<String> allKnownHandles = new ArrayList<>();
+    	
+    	for(Map.Entry<String, Pair<Page, String[]>> entry : pages.entrySet()) {
+			allKnownHandles.add(entry.getValue().getRight()[0]);
 		}
-		if (parentHandle == null) {
-			var errorMessage = "Parent Window cannot be found. Please open a window and restart your test.";
-			log.error(errorMessage);
-			throw new NoSuchWindowException(errorMessage);
+
+		try {
+			FluentWait<WebDriver> wait = new FluentWait<>(driver())
+				       .withTimeout(Time.out().plusSeconds(30))
+				       .pollingEvery(Time.interval())
+				       .ignoring(Exception.class);
+
+			wait.until(d -> {
+				var updatedWindowHandleList = driver().getWindowHandles();
+
+				if(updatedWindowHandleList.isEmpty()) {
+					return false;
+				}
+
+				for(String foundWindowHandle : updatedWindowHandleList) { 
+					if(!allKnownHandles.contains(foundWindowHandle)) {
+						driver().switchTo().window(foundWindowHandle);
+						return true;
+					}
+				}
+
+				return false;
+			});
 		}
-		for (String handle : handles) {
-			if (!handle.equals(parentHandle)) {
-				newHandle = handle;
-			}
+		catch (TimeoutException e) {
+			throw new NoSuchWindowException("Failed to switch to the new window");
```

Comment:
When creating error messages, we should give as much information as possible and assume the person using it won't have access to a stack trace. What page? Why do I care that it wasn't used previously? What do I do now?